src/Feldspar/Compiler/Imperative/ArrayOps.hs

```diff
@@ -167,9 +167,18 @@ arrays e (NativeArray _ t) = [(e,t)]
 arrays e (StructType _ fs) = concat [arrays (StructField e f) t | (f,t) <- fs]
 arrays _ _                 = []
 
+-- | Add types that will also need array operations
+close :: Either Type Type -> [Either Type Type]
+close et
+  | Left  t <- et = map Left  $ t : go t
+  | Right t <- et = map Right $ t : go t
+  where go t | isAwLType t   = t : go (maybe undefined id $ elemTyAwL t)
```

Comment:
There's fromMaybe in Data.Maybe that avoids the id and I don't like error messages from "undefined".

Assuming the undefined is replaced by error, I still don't understand why the code won't trip the error for some array types since go is strict in its first argument.

Reply:
The call to elemTyAwL is guarded by isAwLType, so it will *always* return 'Just t'. fromMaybe semms like a good idea.

Reply:
Found a much better way! Since elemTyAwL returns Nothing if its argument is not an AwL type, I can use that as a pattern guard.